src/main/java/org/apache/aurora/scheduler/configuration/InstanceVariablesSubstitutor.java

```diff
@@ -0,0 +1,135 @@
+  private static final String SUFIX = "}";
+  private static final char ESCAPE = '#';
+  private static final Logger LOG = LoggerFactory.getLogger(InstanceVariablesSubstitutor.class);
+
+  private final StrSubstitutor substitutor;
+  private final ITaskConfig task;
+
+  public static InstanceVariablesSubstitutor getInstance(ITaskConfig taskConfig, int instanceID) {
+    LOG.info(String.format("## Task Instances: %s", taskConfig.getInstances()));
+    if (taskConfig.getInstances().isEmpty()) {
+      return new InstanceVariablesSubstitutor(taskConfig, new StrSubstitutor(StrLookup.noneLookup()));
+    } else {
+      IInstance instance = taskConfig.getInstances().get(instanceID);
+      Map<String, String> variables = instance.getVariables().stream().collect(Collectors.toMap(IVariable::getName,
+              IVariable::getValue));
+
+      StrSubstitutor substitutor = new StrSubstitutor(new StrLookup<String>() {
+        @Override
+        public String lookup(String key) {
+          if (variables.containsKey(key)) {
+            return variables.get(key);
+          } else {
+            throw new IllegalArgumentException(String.format("The variable %s is not present in the task variables %s",
+                    key, variables));
+          }
+        }
+      }, PREFIX, SUFIX, ESCAPE);
+      return new InstanceVariablesSubstitutor(taskConfig, substitutor);
+    }
+  }
+  
+  public InstanceVariablesSubstitutor(ITaskConfig task, StrSubstitutor substitutor) {
+    this.task = task;
+    this.substitutor = substitutor;
+  }
+  
+  /** 
+   * @throws org.apache.aurora.scheduler.configuration.ConfigurationManager.TaskDescriptionException if not all
+   * the instances contains {@link IVariable}s for the specified in the docker parameters or cmdline. 
+   */
+  public static boolean validate(final ITaskConfig taskConfig) throws ConfigurationManager.TaskDescriptionException {
+    
+    final List<Set<String>> variables = Lists.newArrayList();
+    for (IInstance instance : taskConfig.getInstances()) {
+      Set<String> collect = instance.getVariables().stream().map(IVariable::getName).collect(Collectors.toSet());
+      variables.add(collect);
+    }
+    
+    InstanceVariablesSubstitutor in = new InstanceVariablesSubstitutor(taskConfig, new StrSubstitutor(new StrLookup<String>() {
+      @Override public String lookup(String key) {
+        if (variables.isEmpty()) {
+          throw new InstanceVariablesSubstitutorException("No Instances were provided in the job definition but " +
+                  String.format("the variable '%s' was specified.", key));
+        }
+        for (int i = 0; i < variables.size(); i++) {
+          if (!variables.get(i).contains(key)) {
+            throw new InstanceVariablesSubstitutorException(
+              String.format("Instance %s doesn't contain a value for the variable named '%s'", i, key));
+          }
+        }
+        return null;
+      }
+    },PREFIX, SUFIX, ESCAPE));
```

Comment:
I'd rather not fail if a variable wasn't used yet. The validator that invokes this doesn't provide a way to expose warnings, just exceptions that stops the process